data-processing/pdf/process_pdf.py

```diff
@@ -0,0 +1,236 @@
+import requests
+import logging
+from typing import List, Dict
+import sys
+from tempfile import TemporaryDirectory
+import subprocess
+import time
+
+import pdf.grobid_client
+from common.models import BoundingBox as BoundingBoxModel
+from common.models import init_database_connections
+from common.types import CitationData, CitationLocation, SerializableReference, BoundingBox, Matches, Match, Symbol
+from entities.symbols.types import SymbolData, SymbolWithId, SymbolId
+from entities.citations.utils import upload_citations, extract_ngrams, ngram_sim
+from entities.symbols.utils import upload_symbols
+
+class PdfStructureParser(object):
+    def __init__(self, pdf_hash, structure_map):
+        self.pdf_hash = pdf_hash
+        self.structure_map = structure_map
+        self.pages = structure_map['tokens']['pages']
+        self.page_sizes = dict((p['page']['pageNumber'], p['page']) for p in self.pages)
+        self.page_indices = dict((p['page']['pageNumber'], i) for i,p in enumerate(self.pages))
+        self.elements = structure_map['elements']['elementTypes']
+        self.references = None
+
+    def find_token(self, index):
+        for p in self.pages:
+            if (index < len(p['tokens'])):
+                return p['page'], p['tokens'][index]
+            else:
+                index -= len(p['tokens'])
+        return None
+
+
+    def get_text(self,spans):
+        text = []
+        for s in spans:
+            direct = s.get('dehyphenizedText')
+            if direct:
+                text += direct
+            else:
+                for i in range(s['left'], s['right']):
+                    page, token = self.find_token(i)
+                    text.append(token['text'])
+        return ' '.join(text)
+
+    def union(self, bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        if not bbox1:
+            return bbox2
+        x1 = min(bbox1.left,bbox2.left)
+        y1 = min(bbox1.top,bbox2.top)
+        x2 = max(bbox1.left+bbox1.width, bbox2.left+bbox2.width)
+        y2 = max(bbox1.top+bbox1.height, bbox2.top+bbox2.height)
+        return BoundingBoxModel(page = bbox1.page,
```

Comment:
Are pages indexed starting at 0?

Reply:
I admit I'm a little confused here. When I looked at some examples of bounding boxes from the tex pipeline, they seemed to be starting from 1, but I see many bounding boxes with `page=0` in the database. What's the right convention?

Reply:
After refreshing my memory on some of the papers I've run recently, it looks like the right convention is to start at `page=0`. Out of curiosity, which examples are you looking at where they start from 1? I'd like to check on that to make sure the pipeline's being consistent.

Reply:
I used this query to find TeX-pipeline bounding boxes and compare them with the PDF pipeline's:

```
select
       e.source,
       cp.paper_id,
       s.title,
       array_agg(bb.page || '-[' || bb."left" || ',' || bb.top || ',' || bb.width || ',' || bb.height || ']' order by bb.page) as bounding_boxes
from citation c
inner join entity e on e.entity_id = c.id
inner join entityboundingbox ebb on e.id = ebb.entity_id
inner join boundingbox bb on bb.id = ebb.bounding_box_id
inner join citationpaper cp on c.id = cp.citation_id
left outer join summary s on s.paper_id = cp.paper_id
where e.type = 'citation'
and c.paper_id = '3febb2bed8865945e7fddc99efd791887bb7e14f'
group by cp.paper_id, e.source, s.title
order by cp.paper_id, e.source
```

Looking at this again, I think what I was interpreting as 1-index pages were probably false positives on the wrong page.

BTW, I did notice that the schema doesn't have a way to distinguish between entities inserted in different runs, so it's pretty easy to create duplicates.

Reply:
> BTW, I did notice that the schema doesn't have a way to distinguish between entities inserted in different runs, so it's pretty easy to create duplicates.

This is a good point.  The way I've been handling this so far has been to insert data from different runs into different schemas in the same Postgres database. In your scripts, this would mean calling [`init_database_connections`](https://github.com/allenai/scholar-reader/blob/master/data-processing/common/models.py#L148) with a different schema name when you want to insert data into a fresh copy of the tables.

That said, maybe this is a hassle (it has certainly required a specific workflow for me), so let me know if changing the tables to include run version information would be a big value add and it's something I'll add to the docket.